client/state/data-layer/wpcom/sites/comments-tree/index.js

```diff
@@ -35,16 +35,22 @@ export const fetchCommentsTreeForSite = ( { dispatch }, action ) => {
 	);
 };
 
+const mapTree = ( tree, status, type ) => map( tree, comment => ( {
+	commentId: comment[ 0 ],
+	commentParentId: comment[ 2 ],
+	postId: comment[ 1 ],
+	status,
+	type,
+} ) );
+
 export const addCommentsTree = ( { dispatch }, { query }, data ) => {
 	const { siteId, status } = query;
-	const { comments_tree: commentsTree } = data;
 
-	const tree = map( commentsTree, comment => ( {
-		commentId: comment[ 0 ],
-		postId: comment[ 1 ],
-		commentParentId: comment[ 2 ],
-		status,
-	} ) );
+	const tree = [
+		...mapTree( data.comments_tree, status, 'comment' ),
+		...mapTree( data.pingbacks_tree, status, 'pingback' ),
+		...mapTree( data.trackbacks_tree, status, 'trackback' ),
+	];
```

Comment:
this is probably best as-is, but we could also consider another approach…

```js
const tree = [ 'comment', 'pingback', 'trackback' ]
	.map( type => mapTree( data[ `${ type }s_tree` ], status, type ) )
	.reduce( ( final, next ) => final.concat( next ), [] ); // could also be something like ( a, b ) => [ ...a, ...b ]
```

Reply:
You know I've like spent half the day thinking of a pretty functional way to do this (`ary`, `partial`, `flow`, `rearg`, and whatnot), and in the end settled with the simplest approach I could write. 🙂 
With your approach we could drop the `mapTree()` utility and keep all contained in a single expression - which was what I was aiming for anyway!

Reply:
Functional shouldn't mean hard to follow; if it does it's being forced and not applied appropriately.

I see now what's going on: we're wanting to run each item through the shape-converter then join them all into a single list.

We have two typical functions to help us here: `map` and `flatMap` (which runs each item through a collection then flattens the results). In this case it's a little odd just because of the property accesses involved. We could turn this entire thing though into something like this…

```js
flatMap(
	[ 'comment', 'pingback', 'trackback' ],
	type => map(
		data[ `${ type }s_tree` ],
		( [ commentId, … ] ) => ( { … } )
	)
);
```

…but at this point although it's maybe slightly faster I don't find it clearer. Your construction is clear.

Reply:
I like the `flatMap` solution, but the original one, even though it's somewhat repetitive, still looks more readable to me. Since it's unlikely that we'll be adding more types to the tree anytime soon, I'd say stick with the original approach.

Reply:
We might actually add Reblogs at some point, but I've no idea of how they're currently treated and stored. 😄
Though, it would be super easy to add it to the current structure or replace it with the `flatMap`.